File: android/ccda-engine/src/main/kotlin/com/ccdakit/engine/XML/XMLStreamingDocumentBuilder.kt
package com.ccdakit.engine

import org.xmlpull.v1.XmlPullParser

/** Pull-parser helper that captures document metadata and maps section subtrees as they close. */
internal class XMLStreamingDocumentBuilder(
    private val sectionHandler: (CCDAXmlNode, Int) -> CCDASection,
) {
    /** Builds a streamed document from the current parser input. */
    fun build(parser: XmlPullParser): StreamedDocument {
        val headerStack = mutableListOf<CCDAXmlNode>()
        val sectionStack = mutableListOf<CCDAXmlNode>()
        val path = mutableListOf<String>()
        val sections = mutableListOf<CCDASection>()
        var root: CCDAXmlNode? = null
        var skippedBodyDepth = 0

        while (parser.nextToken() != XmlPullParser.END_DOCUMENT) {
            when (parser.eventType) {
                XmlPullParser.START_TAG -> {
                    val node = xmlNode(parser)
                    when {
                        sectionStack.isNotEmpty() -> {
                            sectionStack.last().children.add(node)
                            sectionStack.add(node)
                        }
                        isStructuredBodySectionStart(parser.name, path) -> {
                            sectionStack.add(node)
                        }
                        skippedBodyDepth > 0 -> {
                            skippedBodyDepth += 1
                        }
                        isRootBodyComponent(parser.name, path) -> {
                            skippedBodyDepth = 1
                        }
                        headerStack.isNotEmpty() -> {
                            headerStack.last().children.add(node)
                            headerStack.add(node)
                        }
                        else -> {
                            root = node
                            headerStack.add(node)
                        }
                    }
                    path.add(parser.name)
                }
                XmlPullParser.TEXT, XmlPullParser.CDSECT, XmlPullParser.ENTITY_REF -> {
                    when {
                        sectionStack.isNotEmpty() -> sectionStack.last().textParts.add(parser.text ?: "")
                        skippedBodyDepth == 0 -> headerStack.lastOrNull()?.textParts?.add(parser.text ?: "")
                    }
                }
                XmlPullParser.END_TAG -> {
                    when {
                        sectionStack.isNotEmpty() -> {
                            if (parser.name == CCDAXmlName.SECTION && sectionStack.size == 1) {
                                sections.add(sectionHandler(sectionStack.removeAt(sectionStack.lastIndex), sections.size))
                            } else {
                                sectionStack.removeAt(sectionStack.lastIndex)
                            }
                        }
                        skippedBodyDepth > 0 -> {
                            skippedBodyDepth -= 1
                        }
                        headerStack.isNotEmpty() -> {
                            headerStack.removeAt(headerStack.lastIndex)
                        }
                    }
                    if (path.isNotEmpty()) path.removeAt(path.lastIndex)
                }
            }
        }

        return StreamedDocument(root = root, sections = sections)
    }

    private fun xmlNode(parser: XmlPullParser): CCDAXmlNode =
        CCDAXmlNode(
            name = parser.name,
            attributes = (0 until parser.attributeCount).associate { index ->
                val name = parser.getAttributeName(index)
                val namespace = parser.getAttributeNamespace(index)
                val key = if (
                    namespace == CCDAXmlNamespace.XML_SCHEMA_INSTANCE &&
                    name == CCDAXmlAttribute.TYPE
                ) {
                    CCDAXmlAttribute.XSI_TYPE
                } else {
                    name
                }
                key to parser.getAttributeValue(index)
            },
        )

    private fun isStructuredBodySectionStart(name: String, path: List<String>): Boolean =
        name == CCDAXmlName.SECTION &&
            path.lastOrNull() == CCDAXmlName.COMPONENT &&
            path.dropLast(1).lastOrNull() == CCDAXmlName.STRUCTURED_BODY

    private fun isRootBodyComponent(name: String, path: List<String>): Boolean =
        name == CCDAXmlName.COMPONENT &&
            path.size == 1 &&
            path.firstOrNull() == "ClinicalDocument"
}

/** Stream parse result containing header XML and mapped sections. */
internal data class StreamedDocument(
    val root: CCDAXmlNode?,
    val sections: List<CCDASection>,
)

File: apple/Sources/CCDAEngine/XML/XMLStreamingDocumentBuilder.swift
import Foundation

/// XMLParser delegate that captures document metadata and maps section subtrees as they close.
final class XMLStreamingDocumentBuilder: NSObject, XMLParserDelegate {
    /// Root node containing document-level metadata outside structuredBody.
    private(set) var root: CCDAXMLNode?
    /// Sections mapped in document order.
    private(set) var sections: [CCDASection] = []
    /// First error thrown by a section mapper.
    private(set) var mappingError: Error?

    private var path: [String] = []
    private var headerStack: [CCDAXMLNode] = []
    private var sectionStack: [CCDAXMLNode] = []
    private var skippedBodyDepth = 0
    private let sectionMapper: (CCDAXMLNode, Int) throws -> CCDASection

    /// Creates a streaming builder with a section mapper.
    init(sectionMapper: @escaping (CCDAXMLNode, Int) throws -> CCDASection) {
        self.sectionMapper = sectionMapper
    }

    /// Routes opening elements into either document metadata or the active section subtree.
    func parser(_ parser: XMLParser,
                didStartElement elementName: String,
                namespaceURI: String?,
                qualifiedName qName: String?,
                attributes: [String: String] = [:]) {
        let node = CCDAXMLNode(name: elementName, attributes: attributes)

        if !sectionStack.isEmpty {
            sectionStack.last?.children.append(node)
            sectionStack.append(node)
        } else if isStructuredBodySectionStart(elementName) {
            sectionStack.append(node)
        } else if skippedBodyDepth > 0 {
            skippedBodyDepth += 1
        } else if isRootBodyComponent(elementName) {
            skippedBodyDepth = 1
        } else if let parent = headerStack.last {
            parent.children.append(node)
            headerStack.append(node)
        } else {
            root = node
            headerStack.append(node)
        }

        path.append(elementName)
    }

    /// Appends character content to the active capture target.
    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if !sectionStack.isEmpty {
            sectionStack.last?.text += string
        } else if skippedBodyDepth == 0 {
            headerStack.last?.text += string
        }
    }

    /// Closes elements and maps completed section subtrees immediately.
    func parser(_ parser: XMLParser,
                didEndElement elementName: String,
                namespaceURI: String?,
                qualifiedName: String?) {
        if !sectionStack.isEmpty {
            closeSectionElement(elementName, parser: parser)
        } else if skippedBodyDepth > 0 {
            skippedBodyDepth -= 1
        } else {
            _ = headerStack.popLast()
        }

        _ = path.popLast()
    }

    private func closeSectionElement(_ elementName: String, parser: XMLParser) {
        guard elementName == CCDAXMLElement.section.rawValue, sectionStack.count == 1 else {
            _ = sectionStack.popLast()
            return
        }

        guard let sectionNode = sectionStack.popLast() else { return }

        do {
            let section = try sectionMapper(sectionNode, sections.count)
            sections.append(section)
        } catch {
            mappingError = error
            parser.abortParsing()
        }
    }

    private func isStructuredBodySectionStart(_ elementName: String) -> Bool {
        elementName == CCDAXMLElement.section.rawValue
            && path.last == CCDAXMLElement.component.rawValue
            && path.dropLast().last == CCDAXMLElement.structuredBody.rawValue
    }

    private func isRootBodyComponent(_ elementName: String) -> Bool {
        elementName == CCDAXMLElement.component.rawValue
            && path.count == 1
            && path.first == "ClinicalDocument"
    }
}
